DateTimeUtils.dfy

```diff
@@ -0,0 +1,141 @@
+include "DateTimeConstant.dfy"
+
+module DateTimeUtils {
+  import opened Std.Strings
+  import opened Std.BoundedInts
+  import opened DateTimeConstant
+
+  // Month names for better error messages
+  const MONTH_NAMES: seq<string> := ["January", "February", "March", "April", "May", "June",
+                                     "July", "August", "September", "October", "November", "December"]
+
+  // Function versions for use in function contexts
+  function {:extern "DateTimeImpl", "ToEpochTimeMilliseconds"}
+    {:axiom} ToEpochTimeMillisecondsFunc(year: int32, month: uint8, day: uint8, hour: uint8, minute: uint8, second: uint8, millisecond: uint16): int
+
+  function {:extern "DateTimeImpl", "FromEpochTimeMilliseconds"}
+    {:axiom} FromEpochTimeMillisecondsFunc(epochMillis: int): seq<int32>
+    ensures |FromEpochTimeMillisecondsFunc(epochMillis)| == 7
+    ensures var components := FromEpochTimeMillisecondsFunc(epochMillis);
+            IsValidDateTime(components[0], components[1] as uint8, components[2] as uint8, components[3] as uint8, components[4] as uint8, components[5] as uint8, components[6] as uint16)
+
+  // External function for getting current time components
+  function {:extern "DateTimeImpl", "GetNowComponents"}
```

Comment:
This can't be a `function`, it has to be a `method`. This is because a Dafny `function` has to be a pure deterministic mathematical function that only depends on its inputs, which certainly isn't true here!

This could otherwise cause unsoundness:

```dafny
var t1 := GetNowComponentsFunc();
var t2 := GetNowComponentsFunc();
if t1 != t2 {
    // Dafny believes it's impossible to reach these lines,
    // but it actually is!
    assert false;
    print 1/0;
}
```